helix-core/src/main/java/org/apache/helix/manager/zk/ZKHelixManager.java

```diff
@@ -699,7 +699,7 @@ void createClient() throws Exception {
       try {
         _zkclient.waitUntilConnected(_connectionInitTimeout, TimeUnit.MILLISECONDS);
         handleStateChanged(KeeperState.SyncConnected);
-        handleNewSession();
+        handleNewSession(getSessionId());
```

Comment:
This is dangerous.

_zkclient.waitUntilConnected(_connectionInitTimeout, TimeUnit.MILLISECONDS);
handleStateChanged(KeeperState.SyncConnected);	        handleStateChanged(KeeperState.SyncConnected);
// Session expired and re-created. So next, you are going to create a node with wrong session number.
handleNewSession();	        handleNewSession(getSessionId());

Please lock the whole bock. Or we need to let waitUntilConnected return a session Id synchronizely.

Reply:
Here at the time `getSessionId()`, the value is actually null. We already have `waitUntilConnected()` in `handleNewSession(sessionId)` and `waitUntilConnected()` will update the session id. So my thought is, maybe we can just replace this line `_zkclient.waitUntilConnected(_connectionInitTimeout, TimeUnit.MILLISECONDS);` with the private method `waitUntilConnected()`. This may look cleaner. 
Anyway, I am going to just keep it in this PR, to make this PR a pure new interface change and leave session expiry operation not changed. And I will put the operation change in the next PR.

What do you think?